service/src/main/java/io/camunda/service/UserTaskServices.java

```diff
@@ -40,11 +58,176 @@ public UserTaskServices withAuthentication(final Authentication authentication)
+      variableQueryBuilder =
+          VariableQuery.of(b -> b.filter(variableFilterBuilder.build()).page(searchQueryPage));
+    } else {
+      variableFilterBuilder.processInstanceKeys(
+          new ArrayList<>(scopeKeyMapByProcessInstanceKey.keySet()));
+      variableFilterBuilder.variable(variableValueFilter);
+      variableQueryBuilder =
+          VariableQuery.of(b -> b.filter(variableFilterBuilder.build()).page(searchQueryPage));
+    }
+
+    final var searchResults =
+        variableServices.withAuthentication(authentication).search(variableQueryBuilder);
+
+    final Map<Long, Set<Long>> currentScopeKeyMapByProcessInstanceKey =
+        searchResults.items().stream()
+            .collect(
+                Collectors.groupingBy(
+                    VariableEntity::processInstanceKey,
+                    Collectors.mapping(VariableEntity::scopeKey, Collectors.toSet())));
+
+    if (firstIteration) {
+      initializeScopeKeyMap(
+          scopeKeyMapByProcessInstanceKey, currentScopeKeyMapByProcessInstanceKey);
+    } else {
+      updateScopeKeyMap(scopeKeyMapByProcessInstanceKey, currentScopeKeyMapByProcessInstanceKey);
+    }
+  }
+
+  private void initializeScopeKeyMap(
+      final Map<Long, Set<Long>> scopeKeyMapByProcessInstanceKey,
+      final Map<Long, Set<Long>> currentScopeKeyMapByProcessInstanceKey) {
+    for (final Map.Entry<Long, Set<Long>> entry :
+        currentScopeKeyMapByProcessInstanceKey.entrySet()) {
+      final Long processInstanceId = entry.getKey();
+      final Set<Long> validScopeKeys =
+          entry.getValue().stream()
+              .filter(scopeKey -> !scopeKey.equals(processInstanceId))
+              .collect(Collectors.toSet());
+      if (!validScopeKeys.isEmpty()) {
+        scopeKeyMapByProcessInstanceKey.put(processInstanceId, validScopeKeys);
+      } else {
+        scopeKeyMapByProcessInstanceKey.put(processInstanceId, null);
+      }
+    }
+  }
+
+  private void updateScopeKeyMap(
+      final Map<Long, Set<Long>> scopeKeyMapByProcessInstanceKey,
+      final Map<Long, Set<Long>> currentScopeKeyMapByProcessInstanceKey) {
+    scopeKeyMapByProcessInstanceKey
+        .entrySet()
+        .removeIf(
+            entry -> {
+              final Long processInstanceId = entry.getKey();
+              final Set<Long> validScopeKeys =
+                  entry.getValue() != null
+                      ? entry.getValue().stream()
+                          .filter(Objects::nonNull)
+                          .collect(Collectors.toSet())
+                      : new HashSet<>();
+              final Set<Long> currentScopeKeys =
+                  currentScopeKeyMapByProcessInstanceKey.get(processInstanceId);
+
+              if (currentScopeKeys != null) {
+                validScopeKeys.retainAll(currentScopeKeys);
+              }
+
+              if (entry.getValue() == null || validScopeKeys.isEmpty()) {
+                return currentScopeKeys == null || currentScopeKeys.isEmpty();
```

Comment:
❓ To my understanding, `currentScopeKeys` will always contain at least one key which would be the `processInstanceKey`. Basically, this will never return `true`. Why is this check then necessary?

Reply:
Okay, it is not necessarily the `processInstanceKey`, but IMO it cannot be empty or `null`